### DISCUSSION TITLE
Prepend list items with infinite scroll and keep scroll position

### DISCUSSION THREAD

**Discussion by agnauck** [AvaloniaUI/Avalonia#14388] [Q&A]
I am looking for some help on an implementation for infinite scrolling in both direction, up and down.

The down part is pretty simple. Because when I add new items to the end of the list, then only the scrollbar adjusts automatically. The list stays as is and does not scroll anywhere. This is desired behaviour.

When the list is on top and I prepend new items then by default is scrolls up to the new topmost item. But usually when prepending items in a infinite scrolling scenario the UI should stay as is, and only the scrollbar should adjust slightly to indicate that I can scroll up now.

I tried several scenarios to scroll to the last item prior to the prepend. This brings the item into the view, but usually not as the first item in the list.

Here is a small demo project which I started in case someone wants to play with it and try solutions:
https://github.com/agnauck/AvaloniaListBoxBug3/tree/infinite-scroll

Here is a virtual list component for vue.js. I want to have the same behaviour when prepending new items to my ViewModel which gets displayed using a VirtualizingStackPanel. Item height need to be dynamic which adds more complexity.
https://tangbc.github.io/vue-virtual-scroll-list/#/chat-room

**Comment by agnauck** [comment]
With the current `ScrollIntoView` method we cannot control on where in the view the item gets positioned. The results are also very random.

I think 2 common scenarios are the following:
1) Scroll item into view and position the start of the item at the top of the view

Example: Scroll item 502 into view and position on top
![case-1](https://github.com/AvaloniaUI/Avalonia/assets/1344950/e606e99c-612d-4f75-89cd-4e66f8d8414c)

2) Scroll item into view and position the end of the item at end of the view

Example: Scroll item 503 into view and position at the end
![case-2](https://github.com/AvaloniaUI/Avalonia/assets/1344950/69764779-c0e8-445a-b4b3-201d2417d4fa)

**Reply by sandichhuu** [reply]
I'm facing the same issue.
Cannot make simple chat app if ScrollViewer is not stable.
I known calculate exactly the height/position of scrollbar is expensive task so Avalonia just use estimate.

**Comment by agnauck** [comment]
Did anyone had some success on building a infinite scroll UI?
1) I am still struggling the most with adding items to the top of the list while trying to keep the scroll position
2) The unstable jumping scrollbar with variable items heights. Is there a way to control the scrollbar calculations? I think it could be even better in those cases to just calculate based on items count instead of item height which is not predictable

**Reply by timunie** [reply]
In Avalonia.Labs I have a PR open for a VirtualizationWrapPanel. It can handle different sized items, if you can predict the height. But that isn't straight forward. Logical scrolling is an alternative which has also drawbacks. See Excel rows with random heights. If it exceeds the current viewPort you can't see the bottom. 

What would definitely work is rendering on your own. But this requires also some steep learning curve

**Reply by sandichhuu** [reply]
@agnauck 
I have to fixed the heigh of component inside scroll view.
When changing the height of component after render, they will jumping.
Example: 
I have a chat app, image message should be vertical and horizontal => I make 2 different UI component to handle it.
Place the image holder first to keep the size fixed. (If put empty image, heigh should be 0 and hard to control).
Then just replace downloaded image into.

**Comment by agnauck** [comment]
it looks like most of the issues which were blockers for me here in the past were resolved by the hard work @timunie has done recently. I will proceed any make another POC in the next days with the nightly build.
